### ISSUE TITLE
Cannot use npm install waterline latest version

### ISSUE THREAD

**Issue by golfapipol** [balderdashy/waterline#1365]
**Waterline version**: 0.12.2
**Node version**: 6.2.1
**NPM version**: 3.9.3
**Operating system**: OSX
---
I can't install waterline latest version
npm install waterline

> microtime@0.3.3 install /Users/golfapipol/Desktop/cg/wit-api/node_modules/microtime
> node-gyp rebuild

  CXX(target) Release/obj.target/microtime/src/microtime.o
../src/microtime.cc:10:32: error: unknown type name 'Arguments'; did you mean 'v8::internal::Arguments'?
static Handle<Value> Now(const Arguments &args) {
                               ^~~~~~~~~
                               v8::internal::Arguments

**Comment by sailsbot** [comment]
Hi @golfapipol!  It looks like you may have removed some required elements from the initial comment template, without which I can't verify that this post meets our [contribution guidelines](http://bit.ly/sails-issue-guide). To re-open this issue, please copy the template from [here](https://raw.githubusercontent.com/balderdashy/waterline/master/.github/ISSUE_TEMPLATE), paste it at the beginning of your initial comment, and follow the instructions in the text. Then post a new comment (e.g. "ok, fixed!") so that I know to go back and check.

Sorry to be a hassle, but following these instructions ensures that we can help you in the best way possible and keep the Sails project running smoothly.

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@treeline.io.*

**Comment by calidion** [comment]
confirmed with:
node version: v4.2.3
npm version: 3.5.3
waterline version:  0.1.4

**Comment by calidion** [comment]
maybe waterline should update its microtime package.
```
  CXX(target) Release/obj.target/microtime/src/microtime.o
../src/microtime.cc:10:32: error: ‘Arguments’ does not name a type
 static Handle<Value> Now(const Arguments &args) {
                                ^
../src/microtime.cc:10:43: error: ISO C++ forbids declaration of ‘args’ with no type [-fpermissive]
 static Handle<Value> Now(const Arguments &args) {
                                           ^
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h: In function ‘v8::Handle<v8::Value> Now(const int&)’:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:885:13: error: ‘v8::HandleScope::HandleScope()’ is protected
   V8_INLINE HandleScope() {}
             ^
../src/microtime.cc:11:17: error: within this context
     HandleScope scope;
                 ^
../src/microtime.cc:17:67: warning: ‘v8::Local<v8::Value> node::ErrnoException(int, const char*, const char*, const char*)’ is deprecated (declared at /home/eric/.node-gyp/4.2.3/include/node/node.h:99): Use UVException(isolate, ...) [-Wdeprecated-declarations]
         return ThrowException(ErrnoException(errno, "gettimeofday"));
                                                                   ^
../src/microtime.cc:17:68: error: ‘ThrowException’ was not declared in this scope
         return ThrowException(ErrnoException(errno, "gettimeofday"));
                                                                    ^
../src/microtime.cc:20:18: error: ‘class v8::HandleScope’ has no member named ‘Close’
     return scope.Close(Number::New((t.tv_sec * 1000000.0) + t.tv_usec));
                  ^
../src/microtime.cc:20:70: error: no matching function for call to ‘v8::Number::New(double)’
     return scope.Close(Number::New((t.tv_sec * 1000000.0) + t.tv_usec));
                                                                      ^
../src/microtime.cc:20:70: note: candidate is:
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2486:24: note: static v8::Local<v8::Number> v8::Number::New(v8::Isolate*, double)
   static Local<Number> New(Isolate* isolate, double value);
                        ^
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2486:24: note:   candidate expects 2 arguments, 1 provided
../src/microtime.cc: At global scope:
../src/microtime.cc:23:38: error: ‘Arguments’ does not name a type
 static Handle<Value> NowDouble(const Arguments &args) {
                                      ^
../src/microtime.cc:23:49: error: ISO C++ forbids declaration of ‘args’ with no type [-fpermissive]
 static Handle<Value> NowDouble(const Arguments &args) {
                                                 ^
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h: In function ‘v8::Handle<v8::Value> NowDouble(const int&)’:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:885:13: error: ‘v8::HandleScope::HandleScope()’ is protected
   V8_INLINE HandleScope() {}
             ^
../src/microtime.cc:24:17: error: within this context
     HandleScope scope;
                 ^
../src/microtime.cc:30:67: warning: ‘v8::Local<v8::Value> node::ErrnoException(int, const char*, const char*, const char*)’ is deprecated (declared at /home/eric/.node-gyp/4.2.3/include/node/node.h:99): Use UVException(isolate, ...) [-Wdeprecated-declarations]
         return ThrowException(ErrnoException(errno, "gettimeofday"));
                                                                   ^
../src/microtime.cc:30:68: error: ‘ThrowException’ was not declared in this scope
         return ThrowException(ErrnoException(errno, "gettimeofday"));
                                                                    ^
../src/microtime.cc:33:18: error: ‘class v8::HandleScope’ has no member named ‘Close’
     return scope.Close(Number::New(t.tv_sec + (t.tv_usec * 0.000001)));
                  ^
../src/microtime.cc:33:69: error: no matching function for call to ‘v8::Number::New(double)’
     return scope.Close(Number::New(t.tv_sec + (t.tv_usec * 0.000001)));
                                                                     ^
../src/microtime.cc:33:69: note: candidate is:
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2486:24: note: static v8::Local<v8::Number> v8::Number::New(v8::Isolate*, double)
   static Local<Number> New(Isolate* isolate, double value);
                        ^
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2486:24: note:   candidate expects 2 arguments, 1 provided
../src/microtime.cc: At global scope:
../src/microtime.cc:36:38: error: ‘Arguments’ does not name a type
 static Handle<Value> NowStruct(const Arguments &args) {
                                      ^
../src/microtime.cc:36:49: error: ISO C++ forbids declaration of ‘args’ with no type [-fpermissive]
 static Handle<Value> NowStruct(const Arguments &args) {
                                                 ^
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h: In function ‘v8::Handle<v8::Value> NowStruct(const int&)’:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:885:13: error: ‘v8::HandleScope::HandleScope()’ is protected
   V8_INLINE HandleScope() {}
             ^
../src/microtime.cc:37:17: error: within this context
     HandleScope scope;
                 ^
../src/microtime.cc:43:67: warning: ‘v8::Local<v8::Value> node::ErrnoException(int, const char*, const char*, const char*)’ is deprecated (declared at /home/eric/.node-gyp/4.2.3/include/node/node.h:99): Use UVException(isolate, ...) [-Wdeprecated-declarations]
         return ThrowException(ErrnoException(errno, "gettimeofday"));
                                                                   ^
../src/microtime.cc:43:68: error: ‘ThrowException’ was not declared in this scope
         return ThrowException(ErrnoException(errno, "gettimeofday"));
                                                                    ^
../src/microtime.cc:46:38: error: invalid conversion from ‘int’ to ‘v8::Isolate*’ [-fpermissive]
     Local<Array> array = Array::New(2);
                                      ^
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2965:23: error:   initializing argument 1 of ‘static v8::Local<v8::Array> v8::Array::New(v8::Isolate*, int)’ [-fpermissive]
   static Local<Array> New(Isolate* isolate, int length = 0);
                       ^
../src/microtime.cc:47:30: error: no matching function for call to ‘v8::Integer::New(int)’
     array->Set(Integer::New(0), Uint32::New(t.tv_sec));
                              ^
../src/microtime.cc:47:30: note: candidate is:
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2499:25: note: static v8::Local<v8::Integer> v8::Integer::New(v8::Isolate*, int32_t)
   static Local<Integer> New(Isolate* isolate, int32_t value);
                         ^
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2499:25: note:   candidate expects 2 arguments, 1 provided
../src/microtime.cc:47:53: error: no matching function for call to ‘v8::Uint32::New(__time_t&)’
     array->Set(Integer::New(0), Uint32::New(t.tv_sec));
                                                     ^
../src/microtime.cc:47:53: note: candidate is:
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2499:25: note: static v8::Local<v8::Integer> v8::Integer::New(v8::Isolate*, int32_t)
   static Local<Integer> New(Isolate* isolate, int32_t value);
                         ^
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2499:25: note:   candidate expects 2 arguments, 1 provided
../src/microtime.cc:48:30: error: no matching function for call to ‘v8::Integer::New(int)’
     array->Set(Integer::New(1), Uint32::New(t.tv_usec));
                              ^
../src/microtime.cc:48:30: note: candidate is:
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2499:25: note: static v8::Local<v8::Integer> v8::Integer::New(v8::Isolate*, int32_t)
   static Local<Integer> New(Isolate* isolate, int32_t value);
                         ^
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2499:25: note:   candidate expects 2 arguments, 1 provided
../src/microtime.cc:48:54: error: no matching function for call to ‘v8::Uint32::New(__suseconds_t&)’
     array->Set(Integer::New(1), Uint32::New(t.tv_usec));
                                                      ^
../src/microtime.cc:48:54: note: candidate is:
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2499:25: note: static v8::Local<v8::Integer> v8::Integer::New(v8::Isolate*, int32_t)
   static Local<Integer> New(Isolate* isolate, int32_t value);
                         ^
/home/eric/.node-gyp/4.2.3/include/node/v8.h:2499:25: note:   candidate expects 2 arguments, 1 provided
../src/microtime.cc:50:18: error: ‘class v8::HandleScope’ has no member named ‘Close’
     return scope.Close(array);
                  ^
In file included from ../src/microtime.cc:1:0:
/home/eric/.node-gyp/4.2.3/include/node/v8.h: In function ‘void init(v8::Handle<v8::Object>)’:
/home/eric/.node-gyp/4.2.3/include/node/v8.h:885:13: error: ‘v8::HandleScope::HandleScope()’ is protected
   V8_INLINE HandleScope() {}
             ^
../src/microtime.cc:55:17: error: within this context
     HandleScope scope;
                 ^
../src/microtime.cc:57:39: error: no matching function for call to ‘NODE_SET_METHOD(v8::Handle<v8::Object>&, const char [4], v8::Handle<v8::Value> (&)(const int&))’
     NODE_SET_METHOD(target, "now", Now);
                                       ^
../src/microtime.cc:57:39: note: candidate is:
In file included from ../src/microtime.cc:2:0:
/home/eric/.node-gyp/4.2.3/include/node/node.h:236:13: note: template<class TypeName> void node::NODE_SET_METHOD(const TypeName&, const char*, v8::FunctionCallback)
 inline void NODE_SET_METHOD(const TypeName& recv,
             ^
/home/eric/.node-gyp/4.2.3/include/node/node.h:236:13: note:   template argument deduction/substitution failed:
../src/microtime.cc:57:39: note:   cannot convert ‘Now’ (type ‘v8::Handle<v8::Value>(const int&) {aka v8::Local<v8::Value>(const int&)}’) to type ‘v8::FunctionCallback {aka void (*)(const v8::FunctionCallbackInfo<v8::Value>&)}’
     NODE_SET_METHOD(target, "now", Now);
                                       ^
../src/microtime.cc:58:51: error: no matching function for call to ‘NODE_SET_METHOD(v8::Handle<v8::Object>&, const char [10], v8::Handle<v8::Value> (&)(const int&))’
     NODE_SET_METHOD(target, "nowDouble", NowDouble);
                                                   ^
../src/microtime.cc:58:51: note: candidate is:
In file included from ../src/microtime.cc:2:0:
/home/eric/.node-gyp/4.2.3/include/node/node.h:236:13: note: template<class TypeName> void node::NODE_SET_METHOD(const TypeName&, const char*, v8::FunctionCallback)
 inline void NODE_SET_METHOD(const TypeName& recv,
             ^
/home/eric/.node-gyp/4.2.3/include/node/node.h:236:13: note:   template argument deduction/substitution failed:
../src/microtime.cc:58:51: note:   cannot convert ‘NowDouble’ (type ‘v8::Handle<v8::Value>(const int&) {aka v8::Local<v8::Value>(const int&)}’) to type ‘v8::FunctionCallback {aka void (*)(const v8::FunctionCallbackInfo<v8::Value>&)}’
     NODE_SET_METHOD(target, "nowDouble", NowDouble);
                                                   ^
../src/microtime.cc:59:51: error: no matching function for call to ‘NODE_SET_METHOD(v8::Handle<v8::Object>&, const char [10], v8::Handle<v8::Value> (&)(const int&))’
     NODE_SET_METHOD(target, "nowStruct", NowStruct);
                                                   ^
../src/microtime.cc:59:51: note: candidate is:
In file included from ../src/microtime.cc:2:0:
/home/eric/.node-gyp/4.2.3/include/node/node.h:236:13: note: template<class TypeName> void node::NODE_SET_METHOD(const TypeName&, const char*, v8::FunctionCallback)
 inline void NODE_SET_METHOD(const TypeName& recv,
             ^
/home/eric/.node-gyp/4.2.3/include/node/node.h:236:13: note:   template argument deduction/substitution failed:
../src/microtime.cc:59:51: note:   cannot convert ‘NowStruct’ (type ‘v8::Handle<v8::Value>(const int&) {aka v8::Local<v8::Value>(const int&)}’) to type ‘v8::FunctionCallback {aka void (*)(const v8::FunctionCallbackInfo<v8::Value>&)}’
     NODE_SET_METHOD(target, "nowStruct", NowStruct);
                                                   ^
make: *** [Release/obj.target/microtime/src/microtime.o] 错误 1

```

**Comment by particlebanana** [comment]
microtime isn't a dependency or a child dependency of Waterline as far as I can tell. It could be coming from an adapter or something. Can you post the output of `npm ls`?

**Comment by calidion** [comment]
@particlebanana  please check the version of waterline. 
it is 0.1.4.
not 0.12.4

**Comment by particlebanana** [comment]
@calidion whoa looks like the versioning got way off!

**Comment by particlebanana** [comment]
@calidion I'm not sure what happened, looks like there were 3 publishes all at the same time creating these versions. I've revoked and reset all tokens and set the version back to `0.12`. I'm contacting NPM to see if they have any info. It was from an old npm account.

**Comment by calidion** [comment]
It seems dangerous to release an update on an old version.

Here is a discuss on how npm should be and how to properly tag a version.
https://github.com/npm/npm/issues/13112

**Comment by particlebanana** [comment]
@calidion this was an issue that I have contacted NPM support about. No one on our team published those, they weren't even actual publishes. The latest tag has been reset.

We backport occasional patches to previous versions of Waterline for people who may not be able to upgrade to the latest version. I always manage the `latest` tag to actually be the latest.

In this case something or someone else published these invalid versions. There was nothing malicious on the publishes (they were invalid) but the tag was screwed up. It has since been fixed.
